src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+use {
+    crate::{prelude::*, TimeoutChecker},
+    rand::random,
+    std::{
+        error::Error,
+        fmt,
+        ptr::{read_volatile, write_bytes, write_volatile},
+    },
+};
+
+// TODO: Intend to convert this module to a standalone `no_std` crate
+// TODO: TimeoutChecker will be a trait instead
+
+#[derive(Debug)]
+pub enum MemtestOutcome {
+    Pass,
+    Fail(MemtestFailure),
+}
+
+#[derive(Debug)]
+pub enum MemtestFailure {
+    /// Failure due to the actual value read being different from the expected value
+    UnexpectedValue {
+        address: usize,
+        expected: usize,
+        actual: usize,
+    },
+    /// Failure due to the two memory locations being compared returning two different values
+    /// This is used by tests where memory is splitted in two and random values are written in pairs
+    MismatchedValues {
+        address1: usize,
+        value1: usize,
+        address2: usize,
+        value2: usize,
+    },
+}
+
+#[derive(Debug)]
+pub enum MemtestError {
+    Timeout,
+    Other(anyhow::Error),
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum MemtestType {
+    TestOwnAddress,
+    TestRandomVal,
+    TestXor,
+    TestSub,
+    TestMul,
+    TestDiv,
+    TestOr,
+    TestAnd,
+    TestSeqInc,
+    TestSolidBits,
+    TestCheckerboard,
+    TestBlockSeq,
+}
+
+/// Write the address of each memory location to itself
+/// then read back the value and verify that it matches the expected address
+pub unsafe fn test_own_address(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestOwnAddress");
```

Comment:
Instead of doing this, you can actually just add `#[tracing::instrument]` to the beginning of every function you want logged. It will now automatically log every time it's called.